In [1]:
import pandas as pd


In [4]:
df = pd.read_csv("2024-2025-Data_Librarian_Rabea_Wuttke/Modul_3/data/Library_Usage.csv", low_memory=False, na_values="Null")

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 450359 entries, 0 to 450358
Data columns (total 11 columns):
 #   Column                        Non-Null Count   Dtype  
---  ------                        --------------   -----  
 0   Patron Type Definition        450359 non-null  object 
 1   Total Checkouts               450359 non-null  int64  
 2   Total Renewals                450359 non-null  int64  
 3   Age Range                     449849 non-null  object 
 4   Home Library Definition       450329 non-null  object 
 5   Circulation Active Month      410042 non-null  object 
 6   Circulation Active Year       410042 non-null  float64
 7   Notice Preference Definition  447044 non-null  object 
 8   Provided Email Address        450359 non-null  bool   
 9   Year Patron Registered        450359 non-null  int64  
 10  Within San Francisco County   450138 non-null  object 
dtypes: bool(1), float64(1), int64(3), object(6)
memory usage: 34.8+ MB


## Beschreibung des Datensatz

Es handelt sich um einen Datensatz in dem Ausleihdaten der NutzerInnen der San Francisco Public Library abgebildet sind. Die Grundgesamtheit bilden die aktiven NutzerInnen der Bibliothek, seit 2004 bis 2024. Der Datensatz besteht aus 450359 Stichproben, Merkmalsträger sind die NutzerInnen zu denen bis zu 11 Merkmale (Variablen) erhoben wurden.

In [10]:
df.columns

Index(['Patron Type Definition', 'Total Checkouts', 'Total Renewals',
       'Age Range', 'Home Library Definition', 'Circulation Active Month',
       'Circulation Active Year', 'Notice Preference Definition',
       'Provided Email Address', 'Year Patron Registered',
       'Within San Francisco County'],
      dtype='object')

In [12]:
df['Age Range'] = df['Age Range'].fillna("keine Angaben")

In [14]:
df['Notice Preference Definition'] = df['Notice Preference Definition'].fillna("keine Angaben")

In [16]:
df['Home Library Definition'] = df['Home Library Definition'].fillna("keine Angaben")

In [18]:
df.isna().sum()

Patron Type Definition              0
Total Checkouts                     0
Total Renewals                      0
Age Range                           0
Home Library Definition             0
Circulation Active Month        40317
Circulation Active Year         40317
Notice Preference Definition        0
Provided Email Address              0
Year Patron Registered              0
Within San Francisco County       221
dtype: int64

In [20]:
df_clean = df.dropna()

In [43]:
df_clean.isna().sum()

Patron Type Definition          0
Total Checkouts                 0
Total Renewals                  0
Age Range                       0
Home Library Definition         0
Circulation Active Month        0
Circulation Active Year         0
Notice Preference Definition    0
Provided Email Address          0
Year Patron Registered          0
Within San Francisco County     0
dtype: int64

In [45]:
df_clean

Patron Type Definition  Total Checkouts  Total Renewals  \
0                      Senior                5               0   
1                       Adult                0               0   
2                       Adult                0               0   
3                     Welcome                1               1   
4                      Senior                0               0   
...                       ...              ...             ...   
450133                  Adult                0               0   
450134                  Adult                0               0   
450135                  Adult                0               0   
450136                  Adult                0               0   
450137                  Adult               22               0   

                Age Range Home Library Definition Circulation Active Month  \
0       75 years and over                    Main                      Nov   
1          45 to 54 years                    Main                      Jul   
2          55 to 59 years        Western Addition                      Mar   
3          20 to 24 years                Richmond                      Aug   
4          65 to 74 years                  Sunset                      Mar   
...                   ...                     ...                      ...   
450133     35 to 44 years                    Main                      Jan   
450134     45 to 54 years                    Main                      Mar   
450135     60 to 64 years                    Main                      Dec   
450136     25 to 34 years                    Main                      Jan   
450137     35 to 44 years                  Marina                      Jan   

        Circulation Active Year Notice Preference Definition  \
0                        2022.0                        Email   
1                        2023.0                        Email   
2                        2024.0                        Email   
3                        2022.0                        Email   
4                        2024.0                        Print   
...                         ...                          ...   
450133                   2020.0                        Phone   
450134                   2024.0                        Email   
450135                   2023.0                        Email   
450136                   2021.0                        Email   
450137                   2022.0                        Email   

        Provided Email Address  Year Patron Registered  \
0                         True                    2015   
1                         True                    2019   
2                         True                    2022   
3                         True                    2022   
4                        False                    2023   
...                        ...                     ...   
450133                   False                    2020   
450134                    True                    2017   
450135                    True                    2019   
450136                    True                    2019   
450137                    True                    2019   

       Within San Francisco County  
0                            False  
1                            False  
2                            False  
3                            False  
4                            False  
...                            ...  
450133                       False  
450134                       False  
450135                       False  
450136                       False  
450137                       False  

[409879 rows x 11 columns]

In [47]:
import numpy as np

In [59]:
pd.to_numeric(df_clean['Circulation Active Year'], errors='coerce')

0         2022.0
1         2023.0
2         2024.0
3         2022.0
4         2024.0
           ...  
450133    2020.0
450134    2024.0
450135    2023.0
450136    2021.0
450137    2022.0
Name: Circulation Active Year, Length: 409879, dtype: float64

In [53]:
df_clean['Circulation Active Month'] = pd.to_datetime(
    df_clean['Circulation Active Month'],
    errors='coerce',
    format="%b"
)

/tmp/ipykernel_27318/2685985662.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_clean['Circulation Active Month'] = pd.to_datetime(


In [55]:
df_clean['Circulation Active Month'] = df_clean['Circulation Active Month'].dt.month


/tmp/ipykernel_27318/524060787.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_clean['Circulation Active Month'] = df_clean['Circulation Active Month'].dt.month


In [57]:
df_clean['Membership Duration'] = (df_clean['Circulation Active Year'] - df_clean['Year Patron Registered'])*12 + df_clean['Circulation Active Month']

/tmp/ipykernel_27318/787509261.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_clean['Membership Duration'] = (df_clean['Circulation Active Year'] - df_clean['Year Patron Registered'])*12 + df_clean['Circulation Active Month']


In [61]:
df_clean

Patron Type Definition  Total Checkouts  Total Renewals  \
0                      Senior                5               0   
1                       Adult                0               0   
2                       Adult                0               0   
3                     Welcome                1               1   
4                      Senior                0               0   
...                       ...              ...             ...   
450133                  Adult                0               0   
450134                  Adult                0               0   
450135                  Adult                0               0   
450136                  Adult                0               0   
450137                  Adult               22               0   

                Age Range Home Library Definition  Circulation Active Month  \
0       75 years and over                    Main                        11   
1          45 to 54 years                    Main                         7   
2          55 to 59 years        Western Addition                         3   
3          20 to 24 years                Richmond                         8   
4          65 to 74 years                  Sunset                         3   
...                   ...                     ...                       ...   
450133     35 to 44 years                    Main                         1   
450134     45 to 54 years                    Main                         3   
450135     60 to 64 years                    Main                        12   
450136     25 to 34 years                    Main                         1   
450137     35 to 44 years                  Marina                         1   

        Circulation Active Year Notice Preference Definition  \
0                        2022.0                        Email   
1                        2023.0                        Email   
2                        2024.0                        Email   
3                        2022.0                        Email   
4                        2024.0                        Print   
...                         ...                          ...   
450133                   2020.0                        Phone   
450134                   2024.0                        Email   
450135                   2023.0                        Email   
450136                   2021.0                        Email   
450137                   2022.0                        Email   

        Provided Email Address  Year Patron Registered  \
0                         True                    2015   
1                         True                    2019   
2                         True                    2022   
3                         True                    2022   
4                        False                    2023   
...                        ...                     ...   
450133                   False                    2020   
450134                    True                    2017   
450135                    True                    2019   
450136                    True                    2019   
450137                    True                    2019   

       Within San Francisco County  Membership Duration  
0                            False                 95.0  
1                            False                 55.0  
2                            False                 27.0  
3                            False                  8.0  
4                            False                 15.0  
...                            ...                  ...  
450133                       False                  1.0  
450134                       False                 87.0  
450135                       False                 60.0  
450136                       False                 25.0  
450137                       False                 37.0  

[409879 rows x 12 columns]

In [63]:
df_clean.to_csv("2024-2025-Data_Librarian_Rabea_Wuttke/Modul_3/data/Library_Usage_Clean")